fea-rs/src/compile/tables/var_store.rs

```diff
@@ -0,0 +1,312 @@
+//! Building the ItemVariationStore
+//!
+//! this will probably get moved to write-fonts at some point?
+
+use std::collections::HashMap;
+
+use indexmap::IndexMap;
+use write_fonts::tables::variations::{
+    ItemVariationData, ItemVariationStore, VariationRegion, VariationRegionList,
+};
+
+///
+/// this probably ends up living in write-fonts
+#[derive(Clone, Debug, Default)]
+pub(crate) struct VariationStoreBuilder {
+    // region -> index map
+    all_regions: HashMap<VariationRegion, usize>,
+    // we use an index map so that we have a determenistic ordering,
+    // which lets us write better tests
+    delta_sets: IndexMap<RegionSet, IndexMap<DeltaSet, DeltaKey>>,
+    next_id: DeltaKey,
+}
+
+#[derive(Clone, Debug, Default)]
+pub(crate) struct VariationStoreKeyLookup {
+    map: HashMap<DeltaKey, DeltaKey>,
+}
+
+/// a subset of regions, represented as indicies into the region map
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+struct RegionSet {
+    indicies: Vec<u16>,
+}
+
+#[derive(Clone, Debug, Default, Hash, PartialEq, Eq)]
+struct DeltaSet {
+    values: Vec<i32>,
+}
+
+/// A key that identifies a particular deltaset in an ItemVariationStore.
+///
+/// This corresponds to the outer/inner mapping used to identify deltas.
+/// While constructing a variation store we assign a temporary key to inserted
+/// deltas; these are reordered when the store is built, at which point any
+/// items that reference the store will need to be remapped from their temporary
+/// key to their final one.
```

Comment:
This sounds like a footgun that is all but guarranteed to go off. Can we somehow make it an error to use a temporary key in a final store, ideally at compile time but at runtime would still be better than nothing.

Perhaps we might use different types for var store builder and var store and their respective keys?

Reply:
Unfortunately there's no good way to do this. To explain the problem:

- during the compilation pass, we convert all the fea into `write-fonts` types.
- if we encounter a variable value record or a variable anchor, the variation data is going to get stored in GDEF, and will be referenced by a `VariationIndex` table.
- we don't know the final index until we have collected all the variation data and compiled the `ItemVariationStore`
- So when we initially generate the `ValueRecord` or `Anchor` tables, we assign them temporary Ids
- after we've built the `ItemVariationStore`, we go and remap all of these keys to their final values.

this `DeltaKey` is actually an attempt to do this a bit more carefully (the final type is `VariationIndex`) but at the end of the day we do end up storing temporary values in the tables that we then need to overwrite.

The alternative to this would require us to have an entire duplicate set of types for the layout subtables, which would be thousands of lines of code and very likely cause more problems than it would solve.

There is *one* thing I can think of to hopefully make this easier to debug, which is that we could start the temporary values at 0xffff and count down; that way if we saw a bunch of big values while debugging we would hopefully be better able to diagnose the problem.

Reply:
Update: I woke up with an idea for how we could actually use the type system here: it would require replacing the `VariationIndex` struct (in `write-fonts`) with an enum, that has 'finalized' and 'pending' variants. I don't think this would be much additional code, and would remove any chance of misuse.

I'll open an issue for this.